File: app/src/main/java/com/alex/android_telemetry/telemetry/math/NumericSanitizer.kt
package com.alex.android_telemetry.sensors.normalizer

import kotlin.math.pow

object NumericSanitizer {
    fun raw(x: Double): Double = if (x.isFinite()) x else 0.0

    fun rawOptional(x: Double?): Double? = x?.takeIf { it.isFinite() }

    fun metric(x: Double, digits: Int = 6): Double {
        if (!x.isFinite()) return 0.0
        val scale = 10.0.pow(digits.toDouble())
        return kotlin.math.round(x * scale) / scale
    }

    fun metricOptional(x: Double?, digits: Int = 6): Double? = x?.let { metric(it, digits) }
}

File: ios/Models/NumericSanitizer.swift
import Foundation

enum NumericSanitizer {
    static func raw(_ x: Double) -> Double {
        guard x.isFinite else { return 0.0 }
        return x
    }

    static func rawOptional(_ x: Double?) -> Double? {
        guard let x else { return nil }
        guard x.isFinite else { return nil }
        return x
    }

    static func metric(_ x: Double, digits: Int = 6) -> Double {
        guard x.isFinite else { return 0.0 }
        let scale = pow(10.0, Double(digits))
        return (x * scale).rounded() / scale
    }

    static func metricOptional(_ x: Double?, digits: Int = 6) -> Double? {
        guard let x else { return nil }
        guard x.isFinite else { return nil }
        let scale = pow(10.0, Double(digits))
        return (x * scale).rounded() / scale
    }
}
